### PR TITLE
Allow OverrideOnly method calls for delegation and wrapping in the same class hierarchy

### PR DESCRIPTION
**Pull request by novotnyr**
Allow `OverrideOnly` method calls for delegation and wrapping in the same class hierarchy

For example, do not consider `AnAction#update` as a violation of `OverrideOnly` for delegation and wrapping.

- Introduce bytecode detection for delegate calls with subclasses supported
- Introduce bytecode detection for super calls with subclasses supported
- Introduce `ApiUsageFilter` mechanism as a plug-in method to filter out unnecessary usages for `ApiUsageProcessor`
- Add tests with specific samples of `AnAction`s and `ActionGroup`s.

Please note that this is a more general solution than proposed in the [IDEA-336988/](https://youtrack.jetbrains.com/issue/IDEA-336988/Action-System-add-ApiStatus.OverrideOnly-in-relevant-API), as it supports arbitrary classes and methods.

See [MP-6077](https://youtrack.jetbrains.com/issue/MP-6076/ApiStatus.OverrideOnly-allow-delegation-wrapping)
See [IDEA-336988/](https://youtrack.jetbrains.com/issue/IDEA-336988/Action-System-add-ApiStatus.OverrideOnly-in-relevant-API)

### CODE REVIEW

**Comment by YannCebron** [comment]
Please adjust the title and description of this PR (not `AnAction` specific)

File: intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+fun Method.searchParentOverrides(resolver: Resolver): List<MethodInClass> {
+  return mutableListOf<MethodInClass>().apply {
+    searchParentOverrides(resolver) { klass: ClassFile, method: Method ->
+      add(MethodInClass(method, klass))
+    }
+  }
+}
+
+fun Method.searchParentOverrides(resolver: Resolver, matchHandler: (ClassFile, Method) -> Unit) {
+  var superClassFqn = containingClassFile.superName ?: return
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt:61]: does it skip `java.lang.Object` automatically?

File: intellij-plugin-verifier/verifier-intellij/src/main/java/com/jetbrains/pluginverifier/usages/overrideOnly/OverrideOnlyMethodUsageProcessor.kt [hunk 2]
```diff
@@ -6,37 +6,100 @@ package com.jetbrains.pluginverifier.usages.overrideOnly
...
+
+  private fun Method.isAnnotationPresent(annotationFqn: String, verificationContext: VerificationContext): Boolean {
+    if (findEffectiveMemberAnnotation(annotationFqn, verificationContext.classResolver) != null) {
+      return true
+    }
+
+    val overriddenMethod = searchParentOverrides(verificationContext.classResolver).firstOrNull { (overriddenMethod, c) ->
+       overriddenMethod.findEffectiveMemberAnnotation(annotationFqn, verificationContext.classResolver) != null
+    }
+    return if (overriddenMethod == null) {
```
Comment (YannCebron) [inline intellij-plugin-verifier/verifier-intellij/src/main/java/com/jetbrains/pluginverifier/usages/overrideOnly/OverrideOnlyMethodUsageProcessor.kt:91]: should we use LOG.isLEVELEnabled() here for performance

File: intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt (hunk 1 above)
Comment (novotnyr) [reply intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt:61]: By default, `java.lang.Object` is not skipped. Added a parameter to enable skipping, disabled by default.

Comment (YannCebron) [reply intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt:61]: why would we ever want to check java.lang.Object? it is not annotated with OverrideOnly and never will be

File: intellij-plugin-verifier/verifier-intellij/src/main/java/com/jetbrains/pluginverifier/usages/overrideOnly/OverrideOnlyMethodUsageProcessor.kt (hunk 2 above)
Comment (novotnyr) [reply intellij-plugin-verifier/verifier-intellij/src/main/java/com/jetbrains/pluginverifier/usages/overrideOnly/OverrideOnlyMethodUsageProcessor.kt:91]: Fixed.

File: intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt (hunk 1 above)
Comment (YannCebron) [reply intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt:66]: I don't see a point to keep this parameter

Comment (novotnyr) [reply intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/resolution/Methods.kt:66]: Parameter removed. Documentation added to indicate this behaviour.

**Review by YannCebron — APPROVED** [review]

**Pull request merged**